fms/modules/attention.py

```diff
@@ -21,6 +31,196 @@
+    mask: NotRequired[torch.Tensor]
+    attn_algorithm: NotRequired[str]
+    is_causal_mask: bool
+
+
+def _sdpa_store_op(
+    keys: torch.Tensor,
+    values: torch.Tensor,
+    key_cache: Optional[torch.Tensor],
+    value_cache: Optional[torch.Tensor],
+    **attn_kwargs: Unpack[SDPAAttentionKwargs],
+):
+    keys = keys.transpose(2, 1)
+    values = values.transpose(2, 1)
+
+    if key_cache is not None and value_cache is not None and value_cache.numel() > 0:
+        key_cache_result = torch.cat((key_cache, keys), dim=2)
+        value_cache_result = torch.cat((value_cache, values), dim=2)
+        return (
+            key_cache_result,
+            value_cache_result,
+            key_cache_result,
+            value_cache_result,
+        )
+    else:
+        return (keys, values, keys, values)
+
+
+def _sdpa_compute_op(
+    query: torch.Tensor,
+    key_cache: torch.Tensor,
+    value_cache: torch.Tensor,
+    nheads: int,
+    kvheads: int,
+    p_dropout: float,
+    scale_factor: float,
+    **attn_kwargs: Unpack[SDPAAttentionKwargs],
+):
+    queries = query.transpose(2, 1)
+
+    # no longer transposing prior to store, so need to check this in case of no cache
+    if key_cache.shape[1] != kvheads and key_cache.shape[2] == kvheads:
+        key_cache = key_cache.transpose(2, 1)
+        value_cache = value_cache.transpose(2, 1)
+    mask = attn_kwargs.get("mask", None)
+
+    # Merge rel pos bias and mask into single float mask
+    if mask is not None:
+        # Our expected mask format is bs x q_len x k_len, so to make it broadcastable
+        # we need to create the nheads dimension
+        while len(mask.size()) != 4:  # expects bs (x nheads) x q_len x kv_len
+            mask = mask.unsqueeze(1)
+
+    # Expand kv so black-box attn will work
+    expansion = nheads // kvheads
+    # k/v: b h l d
+    if expansion != 1:
+        keys_e = key_cache.unsqueeze(2).expand(-1, -1, expansion, -1, -1).flatten(1, 2)
+        values_e = (
+            value_cache.unsqueeze(2).expand(-1, -1, expansion, -1, -1).flatten(1, 2)
+        )
+    else:
+        keys_e = key_cache
+        values_e = value_cache
+
+    attn_algorithm = attn_kwargs.get("attn_algorithm", None)
+    if attn_algorithm:
+        # Pick which fused attn kernels will run.
+        use_flash = attn_algorithm == "flash"
+        use_mem_efficient = attn_algorithm == "mem"
+        use_math = attn_algorithm == "math"
+
+        torch.backends.cuda.enable_flash_sdp(use_flash)
+        torch.backends.cuda.enable_mem_efficient_sdp(use_mem_efficient)
+        torch.backends.cuda.enable_math_sdp(use_math)
+
+    attn_mask = mask
+    if attn_mask is not None and attn_mask.dtype != torch.bool:
+        attn_mask = attn_mask.to(dtype=queries.dtype)
+
+    is_causal = attn_kwargs.get(
+        "is_causal_mask",
+        mask is None and not (key_cache.shape[2] != 1 and queries.shape[2] == 1),
+    )
+
+    attn = F.scaled_dot_product_attention(
```

Comment:
add a TODO when we update to 2.7 to use enable_gqa and stop using keys_e and values_e